### PR TITLE
[SPARK-5577] Python udf for DataFrame

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

File: python/pyspark/sql.py [hunk 1]
```diff
@@ -2263,18 +2263,6 @@ def subtract(self, other):
         """
         return DataFrame(getattr(self._jdf, "except")(other._jdf), self.sql_ctx)
 
-    def sample(self, withReplacement, fraction, seed=None):
```
Comment (rxin) [inline python/pyspark/sql.py]: why are we removing sample?

Comment (davies) [inline python/pyspark/sql.py]: there are two sample().

Comment (rxin) [inline python/pyspark/sql.py]: oops mistakenly deleted my own comment

File: python/pyspark/sql.py [hunk 2]
```diff
@@ -2077,9 +2077,9 @@ def dtypes(self):
         """Return all column names and their data types as a list.
 
         >>> df.dtypes
-        [(u'age', 'IntegerType'), (u'name', 'StringType')]
+        [('age', 'integer'), ('name', 'string')]
         """
-        return [(f.name, str(f.dataType)) for f in self.schema().fields]
+        return [(str(f.name), f.dataType.jsonValue()) for f in self.schema().fields]
```
Comment (rxin) [inline python/pyspark/sql.py]: we should use simplestring (which isn't available yet...) we can change it in the future

File: python/pyspark/sql.py [hunk 3]
```diff
@@ -2593,6 +2553,45 @@ def _(col):
     return staticmethod(_)
 
 
+class UserDefinedFunction(object):
+    def __init__(self, func, returnType):
+        self.func = func
+        self.returnType = returnType
+        self._judf = self._create_judf()
+
+    def _create_judf(self):
```
Comment (rxin) [inline python/pyspark/sql.py]: can u add some inline comment explaining what's happening in this function?

File: python/pyspark/sql.py [hunk 4]
```diff
@@ -2651,6 +2650,16 @@ def approxCountDistinct(col, rsd=None):
             jc = sc._jvm.Dsl.approxCountDistinct(_to_java_column(col), rsd)
         return Column(jc)
 
+    @staticmethod
+    def udf(f, returnType=StringType()):
+        """Create a user defined function (UDF)
+
+        >>> slen = Dsl.udf(lambda s: len(s), IntegerType())
+        >>> df.select(slen(df.name).As('slen')).collect()
```
Comment (rxin) [inline python/pyspark/sql.py]: use alias in the docstring

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 5]
```diff
@@ -174,6 +173,22 @@ trait Column extends DataFrame {
   }
 
   /**
+   * Inequality test.
+   * {{{
+   *   // Scala:
+   *   df.select( df("colA") !== df("colB") )
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: you probably want to update the javadoc for !== also.

File: sql/core/src/main/scala/org/apache/spark/sql/Dsl.scala [hunk 6]
```diff
@@ -177,6 +181,23 @@ object Dsl {
     cols.toList.toSeq
   }
 
+  /**
+   * This is a private API for Python
+   * TODO: move this to a private package
+   */
+  def pythonUDF(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dsl.scala]: we already have two functions here - why don't we take this chance to move both into a separate class?

**Comment by davies** [comment]
@rxin I think this is ready.

**Comment by rxin** [comment]
Thanks. Merged.

**Pull request closed without merging**